File: src/YCbCr.js
(function() {
	var global = this;

	/**
	 * @param ybcbr {bytesY, bytesCb, bytesCr, strideY, strideCb, strideCr, width, height, hdec, vdec}
	 * @param TypedArray output: CanvasPixelArray or Uint8ClampedArray to draw RGB into
	 * Assumes that the output array already has alpha channel set to opaque.
	 */
	global.convertYCbCr = function convertYCbCr(ybcbr, output) {
		var width = ybcbr.width,
			height = ybcbr.height,
			hdec = ybcbr.hdec,
			vdec = ybcbr.vdec,
			bytesY = ybcbr.bytesY,
			bytesCb = ybcbr.bytesCb,
			bytesCr = ybcbr.bytesCr,
			strideY = ybcbr.strideY,
			strideCb = ybcbr.strideCb,
			strideCr = ybcbr.strideCr;
		
		if (hdec == vdec == 1) {
			// Optimize for 4:2:0, which is most common
			var outStride = 4 * width,
				outPtr0 = 0,
				outPtr1 = outStride,
				ydec = 0,
				colorCb, colorCr,
				multY00, multY01, multY10, multY11,
				multCrR, multCbCrG, multCbB;
			for (var y = 0; y < height; y += 2) {
				var Y0Ptr = y * strideY,
					Y1Ptr = Y0Ptr + strideY,
					CbPtr = ydec * strideCb,
					CrPtr = ydec * strideCr;
				for (var x = 0; x < width; x += 2) {
					colorCb = bytesCb[CbPtr++];
					colorCr = bytesCr[CrPtr++];

					// Quickie YUV conversion
					// https://en.wikipedia.org/wiki/YCbCr#ITU-R_BT.2020_conversion
					// multiplied by 256 for integer-friendliness
					multCrR = (409 * colorCr) - 57088;
					multCbCrG = (100 * colorCb) + (208 * colorCr) - 34816;
					multCbB = (516 * colorCb) - 70912;
					
					multY00 = (298 * bytesY[Y0Ptr++]);
					output[outPtr0++] = (multY00 + multCrR) >> 8;
					output[outPtr0++] = (multY00 - multCbCrG) >> 8;
					output[outPtr0++] = (multY00 + multCbB) >> 8;
					outPtr0++;

					multY01 = (298 * bytesY[Y0Ptr++]);
					output[outPtr0++] = (multY01 + multCrR) >> 8;
					output[outPtr0++] = (multY01 - multCbCrG) >> 8;
					output[outPtr0++] = (multY01 + multCbB) >> 8;
					outPtr0++;
					
					multY10 = (298 * bytesY[Y1Ptr++]);
					output[outPtr1++] = (multY10 + multCrR) >> 8;
					output[outPtr1++] = (multY10 - multCbCrG) >> 8;
					output[outPtr1++] = (multY10 + multCbB) >> 8;
					outPtr1++;

					multY11 = (298 * bytesY[Y1Ptr++]);
					output[outPtr1++] = (multY11 + multCrR) >> 8;
					output[outPtr1++] = (multY11 - multCbCrG) >> 8;
					output[outPtr1++] = (multY11 + multCbB) >> 8;
					outPtr1++;
				}
				outPtr0 += outStride;
				outPtr1 += outStride;
				ydec++;
			}
		} else {
			for (var y = 0; y < height; y++) {
				var xdec = 0,
					ydec = y >> vdec,
					YPtr = y * strideY,
					CbPtr = ydec * strideCb,
					CrPtr = ydec * strideCr,
					outPtr = y * 4 * width,
					colorY, colorCb, colorCr;
				
				for (var x = 0; x < width; x++) {
					xdec = x >> hdec;
					colorY = bytesY[YPtr + x];
					colorCb = bytesCb[CbPtr + xdec];
					colorCr = bytesCr[CrPtr + xdec];
	
					// Quickie YUV conversion
					// https://en.wikipedia.org/wiki/YCbCr#ITU-R_BT.2020_conversion
					// multiplied by 256 for integer-friendliness
					multY = (298 * colorY);
					output[outPtr++] = (multY + (409 * colorCr) - 57088) >> 8;
					output[outPtr++] = (multY - (100 * colorCb) - (208 * colorCr) + 34816) >> 8;
					output[outPtr++] = (multY + (516 * colorCb) - 70912) >> 8;
					outPtr++;
				}
			}
		}
	}
})();

File: src/YCbCr.c
/**
 * Basic YCbCr->RGB conversion
 *
 * @author Brion Vibber <brion@pobox.com>
 * @copyright 2014
 * @license MIT-style
 */

// Believe it or not these ifs are faster than ?:
// and WAY faster than a function call!
#define CLAMP_FFFF(val) if (val & 0xffff0000) { \
							if (val < 0) { \
								val = 0; \
							} else { \
								val = 0xffff; \
							} \
						}

// This gets repetitive.
#define PIXEL_OUT(ptr) 	r = multY + multCrR; \
						g = multY - multCbCrG; \
						b = multY + multCbB; \
						CLAMP_FFFF(r); \
						CLAMP_FFFF(g); \
						CLAMP_FFFF(b); \
						ptr++; \
						output[ptr++] = r >> 8; \
						output[ptr++] = g >> 8; \
						output[ptr++] = b >> 8;


/**
 * Assumes that the output array already has alpha channel set to opaque.
 */
void convertYCbCr(unsigned char *bytesY, unsigned char *bytesCb, unsigned char *bytesCr,
				  int strideY, int strideCb, int strideCr,
				  int width, int height,
				  int hdec, int vdec,
				  unsigned char *output)
{
	int outStride = 4 * width,
		YPtr = 0, Y0Ptr = 0, Y1Ptr = 0,
		CbPtr = 0, CrPtr = 0,
		outPtr = 0, outPtr0 = 0, outPtr1 = 0,
		colorCb = 0, colorCr = 0,
		multY = 0, multCrR = 0, multCbCrG = 0, multCbB = 0,
		x = 0, y = 0, xdec = 0, ydec = 0,
		r = 0, g = 0, b = 0;

	// Interestingly, the rest of this program after the variable declarations
	// is the same in C and ActionScript. :D
	if (hdec == vdec == 1) {
		// Optimize for 4:2:0, which is most common
		outPtr0 = 0;
		outPtr1 = outStride;
		ydec = 0;
		for (y = 0; y < height; y += 2) {
			Y0Ptr = y * strideY;
			Y1Ptr = Y0Ptr + strideY;
			CbPtr = ydec * strideCb;
			CrPtr = ydec * strideCr;
			for (x = 0; x < width; x += 2) {
				colorCb = bytesCb[CbPtr++];
				colorCr = bytesCr[CrPtr++];

				// Quickie YUV conversion
				// https://en.wikipedia.org/wiki/YCbCr#ITU-R_BT.2020_conversion
				// multiplied by 256 for integer-friendliness
				multCrR   = (409 * colorCr) - 57088;
				multCbCrG = (100 * colorCb) + (208 * colorCr) - 34816;
				multCbB   = (516 * colorCb) - 70912;

				multY = (298 * bytesY[Y0Ptr++]);
				PIXEL_OUT(outPtr0);

				multY = (298 * bytesY[Y0Ptr++]);
				PIXEL_OUT(outPtr0);

				multY = (298 * bytesY[Y1Ptr++]);
				PIXEL_OUT(outPtr1);

				multY = (298 * bytesY[Y1Ptr++]);
				PIXEL_OUT(outPtr1);
			}
			outPtr0 += outStride;
			outPtr1 += outStride;
			ydec++;
		}
	} else {
		outPtr = 0;
		for (y = 0; y < height; y++) {
			xdec = 0;
			ydec = y >> vdec;
			YPtr = y * strideY;
			CbPtr = ydec * strideCb;
			CrPtr = ydec * strideCr;

			for (x = 0; x < width; x++) {
				xdec = x >> hdec;
				colorCb = bytesCb[CbPtr + xdec];
				colorCr = bytesCr[CrPtr + xdec];

				// Quickie YUV conversion
				// https://en.wikipedia.org/wiki/YCbCr#ITU-R_BT.2020_conversion
				// multiplied by 256 for integer-friendliness
				multCrR   = (409 * colorCr) - 57088;
				multCbCrG = (100 * colorCb) + (208 * colorCr) - 34816;
				multCbB   = (516 * colorCb) - 70912;

				multY = 298 * bytesY[YPtr++];
				PIXEL_OUT(outPtr);
			}
		}
	}
}
